### PR TITLE
feat: dynamic rollup configuration for esm flattening

### PR DESCRIPTION
**Pull request by alan-agius4**
## I'm submitting a...

```
[ ] Bug Fix
[X] Feature
[ ] Other (Refactoring, Added tests, Documentation, ...)
```

## Checklist

- [X] Commit Messages follow the [Conventional Commits](https://conventionalcommits.org/) pattern
  - A feature commit message is prefixed "feat:"
  - A bugfix commit message is prefixed "fix:"
- [X] Tests for the changes have been added

## Description
`externals` has been removed in favor of `namedExternals` which is now only used to name the external libraries.

By default all dependencies are now treated as externals and thus are not embedded in the final bundle.

However, If you want to embed one or more of them into the final bundle you can do so by adding the library in the `embeeded` section like so;

```json
{
  "$schema": "../../../src/ng-package.schema.json",
  "lib": {
    "embedded": [
      "lodash",
      "date-fns"
    ]
  }
}
```

While `rollup` will do it's best to guess the name of an external library some modules are hard to guess and it's recommended to provide the name of the external dependency. This can be done by using `namedExternals` in the config section like so;

```json
{
  "$schema": "../../../src/ng-package.schema.json",
  "lib": {
    "namedExternals": {
      "lodash" : "_",
      "date-fns" : "DateFns",
    }
  }
}
```

## Does this PR introduce a breaking change?

```
[X] Yes
[ ] No
```

Closes: #312

### CODE REVIEW

File: src/lib/steps/rollup.ts [hunk 1]
```diff
@@ -23,37 +23,31 @@ export interface RollupOptions {
...
   log.debug(`rollup (v${ROLLUP_VERSION}) ${opts.entry} to ${opts.dest} (${opts.format})`);
-
-  const globals = {
-    // default externals for '@angular/*' and 'rxjs'
-    ...ROLLUP_GLOBALS,
-    // external symbols passed from the user's ng-package.json
-    ...opts.externals,
+  const embedded: string[] = opts.embedded || [];
+  const namedExternals = {
+    'tslib': 'tslib',
```
Comment (alan-agius4) [inline src/lib/steps/rollup.ts]: might also make sense to add some other defaults that are widely used example, `ionic-angular`, `lodash`, `moment`, `date-fns`, `react` etc..

**Comment by alan-agius4** [comment]
Need to update lockfile, though I am on my mobile at the moment!

File: src/lib/steps/rollup.ts [hunk 2]
```diff
@@ -22,38 +21,32 @@ export interface RollupOptions {
...
-        }
+      if (
+        path.isAbsolute(moduleId) ||
+        moduleId.startsWith(".") ||
+        moduleId.startsWith("/") ||
+        moduleId.indexOf("commonjsHelpers") >= 0 || // in case we are embedding a commonjs module we need to include it's helpers also
+        embedded.some(x => x === moduleId)
+      ) {
+        // if it's either 'absolute', marked to embed, starts with a '.' or '/' it's not external
+        return false;
```
Comment (dherges) [inline src/lib/steps/rollup.ts]: What do you think of extracting to a function / lambda?

```ts
const externalModuleIdStrategy = (moduleId: string): boolean => {
  /* ... */
}
```

Then write a unit test for it?

Comment (alan-agius4) [reply src/lib/steps/rollup.ts]: I’d love that, if you tell me where to place the unit test etc... :) I’d find it much more easy to test.

File: src/lib/steps/rollup.ts [hunk 3]
```diff
@@ -77,7 +70,42 @@ export async function rollup(opts: RollupOptions): Promise<void> {
...
+      if (/^rxjs\/symbol/.test(moduleId)) {
+        return "Rx.Symbol";
+      }
+
+      if (/^rxjs\/(add\/)?operator/.test(moduleId)) {
+        return "Rx.Observable.prototype";
+      }
+
+      return namedExternals[moduleId] || ''; // leave it up to rollup to guess the global name
+    },
```
Comment (dherges) [inline src/lib/steps/rollup.ts]: See above. Extract to a function / named lambda and then write a unit test for that piece of functionality?

```ts
const umdModuleIdStrategy = (moduleId: string): string => { /* ... */ }

/* ... */

describe(`umdModuleIdStrategy`, () => {
  it(`should map rxjs/add/operator/<foo> to Rx.Observable.<foo>`, () => {
    /* ... */
  });
});
```

???

Comment (dherges) [reply src/lib/steps/rollup.ts]: Wdyt?

File: src/lib/steps/rollup.ts (hunk 2 above)
Comment (alan-agius4) [reply src/lib/steps/rollup.ts]: Small correction from your example, to test it, you need to export it though!

**Comment by dherges** [comment]
Can you try to reset to `origin/master`? Ideally it should be enough to keep the changes to these files:

screen shot 2017-12-11 at 21 40 48


Unit tests have been added on the master, command: `yarn test:specs`

The test file is named `<being-tested>.spec.ts`. Keep that pattern with `rollup.ts` and `rollup.spec.ts`.

**Comment by dherges** [comment]
Here's an example:

https://github.com/dherges/ng-packagr/blob/0079deb38f64320ab71d02d3f4c0056c5bf08ca9/src/lib/util/ts-transformers.spec.ts#L1-L9

File: src/lib/steps/rollup.ts (hunk 2 above)
Comment (dherges) [reply src/lib/steps/rollup.ts]: Ah, yes! 😄

**Comment by alan-agius4** [comment]
Thanks, will do that.

**Comment by alan-agius4** [comment]
@dherges Unit tests added

**Comment by alan-agius4** [comment]
Bump!

**Comment by dherges** [comment]
From [rollup's API docs](https://rollupjs.org/#big-list-of-options):

> globals: Object of id: name pairs, used for umd/iife bundles.

Let's rename `namedExternals` to`umdModuleIds` since it's only intention is for helping rollup to find umd module ids.

File: src/lib/steps/rollup.spec.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    const FOO_MODULE = 'FooModule';
+    it(`should map 'foo' to '${FOO_MODULE}' when 'namedExternals' is provided`, () => {
+      expect(umdModuleIdStrategy('foo', { 'foo': FOO_MODULE})).to.equal(FOO_MODULE);
+    });
+
+    it(`should map 'foo' to '' when no 'namedExternals' is provided`, () => {
+      expect(umdModuleIdStrategy('foo')).to.empty;
+    });
+  });
+});
```
Comment (dherges) [inline src/lib/steps/rollup.spec.ts]: 👍

File: integration/samples/rollup/README.md [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Sample library: external and embeded dependencies 
+=====================================
```
Comment (dherges) [inline integration/samples/rollup/README.md]: Ok, I think we can move this integration test to `integration/samples/external`

**Comment by dherges** [comment]
@alan-agius4 When you make these two little adjustments, I think it's good to go!

**Comment by alan-agius4** [comment]
@dherges updated as requested.

**Comment by dherges** [comment]
@alan-agius4 perfect!

**Comment by dherges** [comment]
Just added a few words to the readme, hoping to clarify and be precise on the terms "externals", "dependency", "bundle", and so on

**Comment by dherges** [comment]
Thank you Alan, this is a good step to the right direction!

**Comment by alan-agius4** [comment]
Glad to help :)

**Comment by alan-agius4** [comment]
@dherges any other ideas that I can help with? (Sry to write here in the PR but I cant find your email)

**Pull request merged**